internal/scheduler/scheduling/preempting_queue_scheduler.go

```diff
@@ -412,57 +376,70 @@ func (sch *PreemptingQueueScheduler) collectIdsForGangEviction(evictorResult *Ev
 	allGangJobIds := make(map[string]bool)
 	gangNodeIds := make(map[string]bool)
 	seenGangs := make(map[string]bool)
-	for jobId := range evictorResult.EvictedJctxsByJobId {
-		gangId, ok := sch.gangIdByJobId[jobId]
-		if !ok {
+	for _, jctx := range evictorResult.EvictedJctxsByJobId {
+		if !jctx.Job.GetGangInfo().IsGang() {
 			// Not a gang job.
 			continue
 		}
-		if gangId == "" {
-			return nil, nil, errors.Errorf("no gang id found for job %s", jobId)
-		}
+		gangId := jctx.Job.GetGangInfo().Id()
 		if seenGangs[gangId] {
 			// Gang already processed.
 			continue
 		}
 
-		// Look up the gang this job is part of.
-		gangJobIds := sch.jobIdsByGangId[gangId]
-		if len(gangJobIds) == 0 {
+		activeGangJobs, err := sch.getActiveGangJobs(jctx.Job.Queue(), jctx.Job.GetGangInfo().Id())
+		if err != nil {
+			return nil, nil, err
+		}
+		if len(activeGangJobs) == 0 {
 			return nil, nil, errors.Errorf("no jobs found for gang %s", gangId)
 		}
 
 		// Collect all job ids part of that gang.
-		for gangJobId := range gangJobIds {
-			allGangJobIds[gangJobId] = true
-			if nodeId, ok := sch.nodeIdByJobId[gangJobId]; !ok {
-				return nil, nil, errors.Errorf("no node associated with gang job %s", gangJobId)
-			} else if nodeId == "" {
-				return nil, nil, errors.Errorf("empty node id associated with with gang job %s", gangJobId)
-			} else {
-				gangNodeIds[nodeId] = true
-			}
+		for _, gangJob := range activeGangJobs {
+			allGangJobIds[gangJob.Id()] = true
+			gangNodeIds[gangJob.LatestRun().NodeId()] = true
 		}
 		seenGangs[gangId] = true
 	}
 	return allGangJobIds, gangNodeIds, nil
 }
 
+func (sch *PreemptingQueueScheduler) getActiveGangJobs(queue string, gangId string) ([]*jobdb.Job, error) {
```

Comment:
won't be an issue now but I could see this producing some horrible O(n^2) behaviour in the future due to the need to do a liner scan on the gangJobs